src/cli/cli_coap.cpp

```diff
@@ -0,0 +1,356 @@
+/*
+ *  Copyright (c) 2017, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements a simple CLI coap server and client.
+ */
+
+#include <cli/cli.hpp>
+#include <cli/cli_coap.hpp>
+
+namespace Thread {
+namespace Cli {
+
+const CoapCommand Coap::sCommands[] =
+{
+    { "client", &ProcessClient },
+    { "server", &ProcessServer }
+};
+
+Server *Coap::sServer;
+otCoapResource *Coap::sResource;
+otInstance *Coap::sInstance;
+
+void Coap::PrintPayload(otMessage *aMessage)
+{
+    uint8_t payloadLength = otMessageGetLength(aMessage) - otMessageGetOffset(aMessage);
+    uint8_t outputBufferLength = 16;
+    uint8_t bytesPrinted = 0;
+
+    if (payloadLength > 0)
+    {
+        char payload[outputBufferLength] = {};
+        while (bytesPrinted < payloadLength)
+        {
+            otMessageRead(aMessage, otMessageGetOffset(aMessage) + bytesPrinted, &payload, outputBufferLength);
+            bytesPrinted = bytesPrinted + outputBufferLength;
+            sServer->OutputFormat(" payload: %s", payload);
```

Comment:
Depends what you prefer. I copied the OutputBytes function from https://github.com/openthread/openthread/blob/master/src/cli/cli.cpp#L2980 first, but I liked seeing the "text" rather than HEX.